File: tests/Chessie.CSharp.Test/ExtensionsTests.cs
using System;
using Chessie.ErrorHandling;
using Chessie.ErrorHandling.CSharp;
using Microsoft.FSharp.Core;
using NUnit.Framework;

namespace Chessie.CSharp.Test
{
    [TestFixture]
    public class ExtensionsTests
    {
        [Test]
        public void JoinToResultsOfSuccessWorks()
        {
            var result1 = Result<int, string>.Succeed(1, "added one");
            var result2 = Result<int, string>.Succeed(2, "added two");

            var result3 = result1.Join(result2, _ => 0, _ => -0, (i1, i2) => i1 + i2);

            result3.Match(
                ifSuccess: (x, msgs) =>
                {
                    Assert.AreEqual(3, x);
                    Assert.That(msgs, Is.EquivalentTo(new[] { "added one", "added two" }));
                },
                ifFailure: errs => Assert.Fail());
        }

        [Test]
        public void Test()
        {
            Func<Result<string, string>, Result<string, string>, Result<string, string>, Result<string, string>> f = (r1, r2, r3) =>
                from a in r1
                from b in r2
                from c in r3
                select a + b + c;

            f(Result<string, string>.Succeed("1"), Result<string, string>.Succeed("2"), Result<string, string>.Succeed("3")).Match(
                ifSuccess: (s, _) => Assert.That(s, Is.EqualTo("123")),
                ifFailure: _ => Assert.Fail("should not fail"));

            f(Result<string, string>.Succeed("1", "msg1"), Result<string, string>.Succeed("2", "msg2"), Result<string, string>.Succeed("3", "msg3")).Match(
                ifSuccess: (s, list) =>
                {
                    Assert.That(s, Is.EqualTo("123"));
                    Assert.That(list, Is.EquivalentTo(new[] {"msg1", "msg2", "msg3"}));
                },
                ifFailure: list => Assert.Fail("should not fail"));

            f(Result<string, string>.FailWith("fail"), Result<string, string>.Succeed("2"), Result<string, string>.Succeed("3")).Match(
                ifSuccess: (s, _) => Assert.Fail("should fail"),
                ifFailure: list => Assert.That(list, Is.EquivalentTo(new[] { "fail" })));

            f(Result<string, string>.Succeed("1"), Result<string, string>.FailWith("fail"), Result<string, string>.Succeed("3")).Match(
                ifSuccess: (s, _) => Assert.Fail("should fail"),
                ifFailure: list => Assert.That(list, Is.EquivalentTo(new[] { "fail" })));

            f(Result<string, string>.Succeed("1"), Result<string, string>.FailWith("fail1"), Result<string, string>.FailWith("fail2")).Match(
                ifSuccess: (s, _) => Assert.Fail("should fail"),
                ifFailure: list => Assert.That(list, Is.EquivalentTo(new[] { "fail1" })));
        }
        
        [Test]
        public void ToResultOnSomeShouldSucceed()
        {
            var opt = FSharpOption<int>.Some(42);
            var result = opt.ToResult("error");
            result.Match(
                ifSuccess: (x, msgs) =>
                {
                    Assert.AreEqual(42, x);
                    Assert.That(msgs, Is.Empty);
                },
                ifFailure: errs => Assert.Fail());
        }

        [Test]
        public void ToResultOnNoneShoulFail()
        {
            var opt = FSharpOption<int>.None;
            var result = opt.ToResult("error");
            result.Match(
                ifSuccess: (x, _) => Assert.Fail(),
                ifFailure: errs => Assert.That(errs, Is.EquivalentTo(new[] {"error"})));
        }
    }
}

File: tests/Chessie.VisualBasic.Tests/ExtensionsTests.vb
#If Not __MonoCS__ Then
Imports Chessie.ErrorHandling
Imports Chessie.ErrorHandling.CSharp
Imports Microsoft.FSharp.Core
Imports NUnit.Framework

<TestFixture>
Public Class ExtensionsTests
    <Test>
    Public Sub JoinToResultsOfSuccessWorks()
        Dim result1 = Result(Of Integer, String).Succeed(1, "added one")
        Dim result2 = Result(Of Integer, String).Succeed(2, "added two")

        Dim result3 = result1.Join(result2, Function(i) 0, Function(i) (-0), Function(i1, i2) i1 + i2)

        result3.Match(Sub(x, msgs)
                          Assert.AreEqual(3, x)
                          Assert.That(msgs, [Is].EquivalentTo({"added one", "added two"}))
                      End Sub,
                      Sub(errs) Assert.Fail())
    End Sub

    <Test>
    Public Sub Test()
        Dim f As Func(Of Result(Of String, String), Result(Of String, String), Result(Of String, String), Result(Of String, String)) = Function(r1, r2, r3) From a In r1 From b In r2 From c In r3 Select a + b + c

        f(Result(Of String, String).Succeed("1"), Result(Of String, String).Succeed("2"), Result(Of String, String).Succeed("3")).Match(Sub(s, list) Assert.That(s, [Is].EqualTo("123")),
                                                                                                                                        Sub(list) Assert.Fail("should not fail"))

        f(Result(Of String, String).Succeed("1", "msg1"), Result(Of String, String).Succeed("2", "msg2"), Result(Of String, String).Succeed("3", "msg3")).Match(Sub(s, list)
                                                                                                                                                                    Assert.That(s, [Is].EqualTo("123"))
                                                                                                                                                                    Assert.That(list, [Is].EquivalentTo({"msg1", "msg2", "msg3"}))
                                                                                                                                                                End Sub,
                                                                                                                                                                Sub(list) Assert.Fail("should not fail"))

        f(Result(Of String, String).FailWith("fail"), Result(Of String, String).Succeed("2"), Result(Of String, String).Succeed("3")).Match(Sub(s, list) Assert.Fail("should fail"),
                                                                                                                                            Sub(list) Assert.That(list, [Is].EquivalentTo({"fail"})))

        f(Result(Of String, String).Succeed("1"), Result(Of String, String).FailWith("fail"), Result(Of String, String).Succeed("3")).Match(Sub(s, list) Assert.Fail("should fail"),
                                                                                                                                            Sub(list) Assert.That(list, [Is].EquivalentTo({"fail"})))

        f(Result(Of String, String).Succeed("1"), Result(Of String, String).FailWith("fail1"), Result(Of String, String).FailWith("fail2")).Match(Sub(s, list) Assert.Fail("should fail"),
                                                                                                                                                  Sub(list) Assert.That(list, [Is].EquivalentTo({"fail1"})))
    End Sub

    <Test>
    Public Sub ToResultOnSomeShouldSucceed()
        Dim opt = FSharpOption(Of Integer).Some(42)
        Dim result = opt.ToResult("error")
        result.Match(Sub(x, msgs)
                         Assert.AreEqual(42, x)
                         Assert.That(msgs, [Is].Empty)
                     End Sub,
                     Sub(errs) Assert.Fail())
    End Sub

    <Test>
    Public Sub ToResultOnNoneShoulFail()
        Dim opt = FSharpOption(Of Integer).None
        Dim result = opt.ToResult("error")
        result.Match(Sub(x, list) Assert.Fail(),
                     Sub(errs) Assert.That(errs, [Is].EquivalentTo({"error"})))
    End Sub

    <Test>
    Public Sub MapFailureOnSuccessShouldReturnSuccess()
        Result(Of Integer, String).Succeed(42, "warn1").MapFailure(Function(list) {42}).Match(Sub(v, msgs)
                                                                                                  Assert.AreEqual(42, v)
                                                                                                  Assert.That(msgs, [Is].Empty)
                                                                                              End Sub,
                                                                                              Sub(errs) Assert.Fail())
    End Sub

    <Test>
    Public Sub MapFailureOnFailureShouldMapOverError()
        Result(Of Integer, String).FailWith({"err1", "err2"}).MapFailure(Function(list) {42}).Match(Sub(v, msgs) Assert.Fail(),
                                                                                                    Sub(errs) Assert.That(errs, [Is].EquivalentTo({42})))
    End Sub

    <Test>
    Public Sub MapFailureOnFailureShouldMapOverListOfErrors()
        Result(Of Integer, String).FailWith({"err1", "err2"}).MapFailure(Function(errs) errs.[Select](Function(err)
                                                                                                          Select Case err
                                                                                                              Case "err1"
                                                                                                                  Return 42
                                                                                                              Case "err2"
                                                                                                                  Return 43
                                                                                                              Case Else
                                                                                                                  Return 0
                                                                                                          End Select
                                                                                                      End Function)).Match(Sub(v, msgs) Assert.Fail(),
                                                                                                                           Sub(errs) Assert.That(errs, [Is].EquivalentTo({42, 43})))
    End Sub
End Class
#End If